nx/lib/nx/binary_backend.ex

```diff
@@ -1082,6 +1083,186 @@ defmodule Nx.BinaryBackend do
+                            acc
+                            |> Map.put(:d_qp, x)
+                            |> Map.put(:row_q, acc.row_q <> x)
+
+                          {^q, _} ->
+                            Map.put(acc, :row_q, acc.row_q <> x)
+
+                          {row, _} when row < p ->
+                            Map.put(acc, :before_row_p, acc.before_row_p <> x)
+
+                          {row, _} when row > p and row < q ->
+                            Map.put(acc, :before_row_p, acc.between_rows_p_and_q <> x)
+
+                          _ ->
+                            Map.put(acc, :after_row_q, acc.after_row_q <> x)
+                        end
+
+                      {row, col} =
+                        if col + 1 == n do
+                          {row + 1, 0}
+                        else
+                          {row, col + 1}
+                        end
+
+                      acc |> Map.put(:row, row) |> Map.put(:col, col)
+                  end
+                end
+
+              {rot_l, rot_l_t, rot_r, shape_l, shape_l_t, shape_r} =
+                jacobi_rotators(d_pp, d_pq, d_qp, d_qq)
+
+              {_, num_bits} = output_type
+              row_size = num_bits * n
+
+              <<row_p::bitstring-size(row_size), row_q::bitstring>> =
+                dot_binary(
+                  row_l_t,
+                  shape_l_t,
+                  output_type,
+                  d_values.row_p <> d_values.row_q,
+                  {2, n},
+                  output_type,
+                  output_type
+                )
+
+              # D[[p, q], :] = [[row_p], [row_q]]
+              d_bin =
+                d_values.before_row_p <>
+                  row_p <> d_values.between_rows_p_and_q <> row_q <> d_values.after_row_q
+
+              d_cols_pq = aggregate_axes(d_bin, [0..(m - 1), [p, q]], {m, n}, num_bits)
+
+              d_cols_pq =
+                dot_binary(
+                  d_cols_pq,
+                  {m, 2},
+                  output_type,
+                  rot_r,
+                  shape_r,
+                  output_type,
+                  output_type
+                )
+
+              u_cols_pq = aggregate_axes(u_bin, [0..(m - 1), [p, q]], {m, m}, num_bits)
+
+              u_cols_pq =
+                dot_binary(
+                  u_cols_pq,
+                  {m, 2},
+                  output_type,
+                  rot_l,
+                  shape_l,
+                  output_type,
+                  output_type
+                )
+
+              v_cols_pq = aggregate_axes(v_bin, [0..(n - 1), [p, q]], {n, n}, num_bits)
+
+              v_cols_pq =
+                dot_binary(
+                  v_cols_pq,
+                  {n, 2},
+                  output_type,
+                  rot_r,
+                  shape_r,
+                  output_type,
+                  output_type
+                )
+
+              # assign d_cols_pq to d_bin
```

Comment:
@josevalim @seanmor5 I got a little confused in this last part of the algorithm. Basically, I need a way to translate:
`d[:, [p, q]] = np.matmul(d[:, [p, q]], rot_t)`

If these were rows, I could use the same trick as before where I can slice the bitstrings out and replace them, but I'm not entirely sure if there's an easier way of doing this assignment other than something like:

```elixir
for <<match!(x, 0) <- d_bin>>, reduce: {"", 0, 0, d_bin_cols_pq} do
{result, row, col, <<match!(h, 0), tail :: bitstring>>} when col in [p, q] -> 
  # inc row and col accordingly
  {result <> h, row, col, tail}
{result, row, col, data} ->
 # inc row and col accordingly
 {result <> x, row, col, data}
end
```

Copying this left and right seems a little iffy, but maybe this can become an useful util for this module?

Reply:
I have dropped a comment which may be related to this. Since we are talking only about matrices, it may be more convenient for us to redesign some of those operations to work with lists (or list of lists or list of tuples), it will likely be simpler than working with binaries on every step along the way.

If you find it more comfortable, you can even start by modifying QR first - there is a dot_binary in there which likely can work on lists, and then bring this up.